modin/core/dataframe/pandas/utils.py

```diff
@@ -44,3 +47,34 @@ def concatenate(dfs):
                 i, pandas.Categorical(df.iloc[:, i], categories=union.categories)
             )
     return pandas.concat(dfs)
+
+
+def merge_partitioning(left, right, axis=1):
```

Comment:
* Why is such a partitioning scheme better than the default one?
* Is it suitable only for merge operation?
* Is it beneficial to just make this behavior the default? (when `num_splits=auto`)

Reply:
> Why is such a partitioning scheme better than the default one?

What do you mean by 'default'? 

If 'default' is the current behavior then the answer would be is that it benefits from knowing in advance an approximate shape of the result. By default, the number of splits across an axis is equal to `NPartitions` (let's say it's 112 for example :D), meaning that it's highly likely that there'll be A LOT of empty partitions that will then slow down overall performance.

> Is it suitable only for merge operation?

Yes. `.merge` and `.join` to be precise. We don't have other QC methods that insert 2D frames bypassing `modin_frame.concat`. There are also `.wrtie_items` and `.setitems` but they seem to not change the shape of the frame.

> Is it beneficial to just make this behavior the default? (when num_splits=auto)

Inside `.apply_full_axis` we don't know the shape of `other` frame we're inserting into the source one.

Reply:
> Inside .apply_full_axis we don't know the shape of other frame we're inserting into the source one.

What about `broadcast_apply_full_axis`?

Reply:
well, it's not generally guaranteed that we're going to insert `other` into `self`. There are a lot of use cases where `other` is used just to broadcast a certain subset of columns of the same frame to every partition (thus the resulting shape may not grow at all). 

I think it's reasonable to parametrize this behavior using the `num_splits` parameter. After all, this is quite internal functionality so it'll be used by developers who should understand all the benefits & drawbacks of preserving/not preserving/specifying a custom number of splits for a particular operation. I don't think we could tune partitioning automatically for every single operation, we mostly need to have a hint passed as a parameter by a developer.